internal/xds/translator/basicauth.go

```diff
@@ -31,97 +32,81 @@ type basicAuth struct {
 
 var _ httpFilter = &basicAuth{}
 
-// patchHCM builds and appends the basic_auth Filters to the HTTP Connection Manager
+// patchHCM builds and appends the basic_auth Filter to the HTTP Connection Manager
 // if applicable, and it does not already exist.
-// Note: this method creates an basic_auth filter for each route that contains an BasicAuth config.
-// The filter is disabled by default. It is enabled on the route level.
 func (*basicAuth) patchHCM(mgr *hcmv3.HttpConnectionManager, irListener *ir.HTTPListener) error {
-	var errs error
-
 	if mgr == nil {
 		return errors.New("hcm is nil")
 	}
-
 	if irListener == nil {
 		return errors.New("ir listener is nil")
 	}
+	if hcmContainsFilter(mgr, basicAuthFilter) {
+		return nil
+	}
 
-	for _, route := range irListener.Routes {
-		if !routeContainsBasicAuth(route) {
-			continue
-		}
-
-		// Only generates one BasicAuth Envoy filter for each unique name.
-		// For example, if there are two routes under the same gateway with the
-		// same BasicAuth config, only one BasicAuth filter will be generated.
-		if hcmContainsFilter(mgr, basicAuthFilterName(route.Security.BasicAuth)) {
-			continue
-		}
+	var (
+		irBasicAuth *ir.BasicAuth
+		filter      *hcmv3.HttpFilter
+		err         error
+	)
 
-		filter, err := buildHCMBasicAuthFilter(route.Security.BasicAuth)
-		if err != nil {
-			errs = errors.Join(errs, err)
-			continue
+	for _, route := range irListener.Routes {
+		if route.Security != nil && route.Security.BasicAuth != nil {
+			irBasicAuth = route.Security.BasicAuth
+			break
 		}
-
-		mgr.HttpFilters = append(mgr.HttpFilters, filter)
+	}
+	if irBasicAuth == nil {
+		return nil
 	}
 
-	return errs
+	// We use the first route that contains the basicAuth config to build the filter.
+	// The HCM-level filter config doesn't matter since it is overwritten at the route level.
+	if filter, err = buildHCMBasicAuthFilter(irBasicAuth); err != nil {
+		return err
+	}
+	mgr.HttpFilters = append(mgr.HttpFilters, filter)
+	return err
 }
 
 // buildHCMBasicAuthFilter returns a basic_auth HTTP filter from the provided IR HTTPRoute.
 func buildHCMBasicAuthFilter(basicAuth *ir.BasicAuth) (*hcmv3.HttpFilter, error) {
-	basicAuthProto := basicAuthConfig(basicAuth)
+	var (
+		basicAuthProto *basicauthv3.BasicAuth
+		basicAuthAny   *anypb.Any
+		err            error
+	)
 
-	if err := basicAuthProto.ValidateAll(); err != nil {
+	basicAuthProto = &basicauthv3.BasicAuth{
+		Users: &corev3.DataSource{
+			Specifier: &corev3.DataSource_InlineBytes{
+				InlineBytes: basicAuth.Users,
```

Comment:
Are we using the first route's security policy basic config to define the top-level HCM "global" config? It's overriden on each route, but makes XDS look a bit strange... Maybe just leave it blank?

Reply:
We can't leave it blank because the `Users` field is mandatory.